File: 01-Programming-Languages/TypeScript/Data-Structures/Heap/kthLargest.ts
/*
Problem: Kth Largest Element in a Stream
Description: Find kth largest element using a min-heap of size k.

Approach:
- Maintain min-heap of size k; root is the kth largest

Time Complexity: O(n log k)
Space Complexity: O(k)

Example:
Input: stream = [4,5,8,2], k = 3, add(3), add(5), add(10), add(9), add(4)
Output: 4, 5, 5, 8, 8
*/

class MinHeap {
  private heap: number[] = [];

  size(): number { return this.heap.length; }

  peek(): number { return this.heap[0]; }

  push(val: number): void {
    this.heap.push(val);
    this._bubbleUp(this.heap.length - 1);
  }

  pop(): number {
    const top = this.heap[0];
    const bottom = this.heap.pop()!;
    if (this.heap.length > 0) {
      this.heap[0] = bottom;
      this._sinkDown(0);
    }
    return top;
  }

  private _bubbleUp(idx: number): void {
    while (idx > 0) {
      const parent = Math.floor((idx - 1) / 2);
      if (this.heap[parent] <= this.heap[idx]) break;
      [this.heap[parent], this.heap[idx]] = [this.heap[idx], this.heap[parent]];
      idx = parent;
    }
  }

  private _sinkDown(idx: number): void {
    const length = this.heap.length;
    while (true) {
      let smallest = idx;
      const left = 2 * idx + 1;
      const right = 2 * idx + 2;
      if (left < length && this.heap[left] < this.heap[smallest]) smallest = left;
      if (right < length && this.heap[right] < this.heap[smallest]) smallest = right;
      if (smallest === idx) break;
      [this.heap[idx], this.heap[smallest]] = [this.heap[smallest], this.heap[idx]];
      idx = smallest;
    }
  }
}

class KthLargest {
  private k: number;
  private heap: MinHeap;

  constructor(k: number, nums: number[]) {
    this.k = k;
    this.heap = new MinHeap();
    for (const num of nums) this.add(num);
  }

  add(val: number): number {
    this.heap.push(val);
    if (this.heap.size() > this.k) this.heap.pop();
    return this.heap.peek();
  }
}

const kth = new KthLargest(3, [4, 5, 8, 2]);
console.log('add(3):', kth.add(3));
console.log('add(5):', kth.add(5));
console.log('add(10):', kth.add(10));
console.log('add(9):', kth.add(9));
console.log('add(4):', kth.add(4));

File: 01-Programming-Languages/C/Data-Structures/Heap/kth_largest.c
/*
Problem: Kth Largest Element (Min-Heap)
Find the kth largest element in an array using a min-heap of size k.

Approach:
- Maintain min-heap of size k, insert elements, keep smallest of top k

Time Complexity: O(n log k)
Space Complexity: O(k)

Example:
Input: [3, 2, 1, 5, 6, 4], k = 2
Output: 5
*/

#include <stdio.h>
#include <stdlib.h>

typedef struct
{
    int *data;
    int size;
    int cap;
} MinHeap;

MinHeap *create_heap(int cap)
{
    MinHeap *h = (MinHeap *)malloc(sizeof(MinHeap));
    h->data = (int *)malloc(cap * sizeof(int));
    h->size = 0;
    h->cap = cap;
    return h;
}

void swap(int *a, int *b) { int t = *a; *a = *b; *b = t; }

void heapify(MinHeap *h, int i)
{
    int smallest = i, left = 2 * i + 1, right = 2 * i + 2;
    if (left < h->size && h->data[left] < h->data[smallest]) smallest = left;
    if (right < h->size && h->data[right] < h->data[smallest]) smallest = right;
    if (smallest != i) { swap(&h->data[i], &h->data[smallest]); heapify(h, smallest); }
}

void push(MinHeap *h, int val)
{
    if (h->size < h->cap) { h->data[h->size++] = val; }
    else if (val > h->data[0]) { h->data[0] = val; }
    for (int i = h->size / 2 - 1; i >= 0; i--) heapify(h, i);
    if (h->size == h->cap) heapify(h, 0);
}

int peek(MinHeap *h) { return h->data[0]; }

int find_kth_largest(int arr[], int n, int k)
{
    MinHeap *h = create_heap(k);
    for (int i = 0; i < n; i++) push(h, arr[i]);
    int res = peek(h);
    free(h->data);
    free(h);
    return res;
}

int main()
{
    int arr[] = {3, 2, 1, 5, 6, 4};
    int n = sizeof(arr) / sizeof(arr[0]);
    int k = 2;
    printf("Kth largest (k=%d): %d\n", k, find_kth_largest(arr, n, k));
    return 0;
}
